### PR TITLE
[FIX][SCHEMA] Add conditionals for PET ReconMethod* and ReconFilter

### PR DESCRIPTION
**Pull request by bendhouseart**
Adds conditionals to change required PET fields to optional when "parent" fields are unknown or "none", specifically:

* add conditionals for ReconMethod Metadata

* add min and max values to scatter fraction

* add conditional for ReconFilterSize

Intention is to avoid users adding unnecessary place holders to pass validation, e.g. [0, 0], ["none", "none"], hopefully avoids ambiguity instead of increasing it.

Arguments could be made that fields should always be required, but several recent use cases have prompted this minor change to the spec. Additionally, it's more straight forward (to this user) to toggle optional than to define `null` placeholders for values and datatypes.

### CODE REVIEW

File: src/schema/objects/metadata.yaml [hunk 1]
```diff
@@ -2660,6 +2660,7 @@ ScatterFraction:
     type: number
     minimum: 0
     maximum: 100
+
```
Comment (tsalo) [inline src/schema/objects/metadata.yaml:2663]: Is this new line accidental?

File: src/schema/rules/sidecars/pet.yaml [hunk 2]
```diff
@@ -176,6 +180,25 @@ PETReconstruction:
     SinglesRate: recommended
     RandomRate: recommended
 
+# set required reconstruction fields to optional when "none" is provided as a parameter label
+EntitiesReconMethodMetadata:
+  selectors:
+    - modality == "pet"
+    - suffix == "pet"
+    - count(sidecar.ReconMethodParameterLabels, "none") >= 1
```
Comment (tsalo) [inline src/schema/rules/sidecars/pet.yaml:188]:
```suggestion
    - "none" in sidecar.ReconMethodParameterLabels
```

I _think_ this is the preferred way of saying "X contains 'string'"

File: src/schema/rules/sidecars/pet.yaml [hunk 3]
```diff
@@ -150,12 +150,16 @@ PETReconstruction:
       description_addendum: This partly matches the DICOM Tag (0054,1103) `Reconstruction Method`.
     ReconMethodParameterValues:
       level: required
+      level_addendum: |
+        Required if `ReconMethodParameterLabels` does not contain 'none', optional if
+        no labels exist.
```
Comment (tsalo) [inline src/schema/rules/sidecars/pet.yaml:155]: I think this needs to be reflected in the selectors- i.e., we need two mutually exclusive sets of selectors. One with `"none" in sidecar.ReconMethodParameterLabels` and one with `!("none" in sidecar.ReconMethodParameterLabels)`.

Comment (tsalo) [reply src/schema/rules/sidecars/pet.yaml:155]: Also, shouldn't the level_addendum be applied to `ReconMethodParameterUnits` as well?

File: src/schema/rules/sidecars/pet.yaml [hunk 4]
```diff
@@ -150,12 +150,16 @@ PETReconstruction:
...
       description_addendum: This partly matches the DICOM Tag (0054,1103) `Reconstruction Method`.
     ReconFilterType:
       level: required
       description_addendum: This partly matches the DICOM Tag (0018,1210) `Convolution Kernel`.
     ReconFilterSize:
       level: required
+      level_addendum: Required if `ReconFilterType` is not 'none'.
       description_addendum: This partly matches the DICOM Tag (0018,1210) `Convolution Kernel`.
```
Comment (tsalo) [inline src/schema/rules/sidecars/pet.yaml:163]: I think this needs to be split out into a separate rule as well.

**Review by tsalo — DISMISSED** [review]
I think the rules need to be reconfigured a bit.

File: src/schema/objects/metadata.yaml (hunk 1 above)
Comment (bendhouseart) [inline src/schema/objects/metadata.yaml:2663]:
```suggestion
```

Comment (bendhouseart) [reply src/schema/objects/metadata.yaml:2663]: Yes.

File: src/schema/rules/sidecars/pet.yaml (hunk 2 above)
Comment (bendhouseart) [reply src/schema/rules/sidecars/pet.yaml:188]: I had originally done it that way, but changed it after reading over Chris' updated schema readme. It wasn't clear to me whether or not a an element in an array was a subfield or not. I felt that using count made it much more clear that "none" was an element in an array. Perhaps I just need to update the documentation clarify that the `in` operator works this way on arrays.

File: src/schema/rules/sidecars/pet.yaml (hunk 4 above)
Comment (bendhouseart) [reply src/schema/rules/sidecars/pet.yaml:163]: I'm sorry, I don't understand what you mean by that. Is that not what is defined below in `EntitiesReconFilterData`?

File: src/schema/rules/sidecars/pet.yaml (hunk 3 above)
Comment (bendhouseart) [reply src/schema/rules/sidecars/pet.yaml:155]: Fits with the logic, I've never seen them have units so I don't pay them much mind.

Comment (bendhouseart) [reply src/schema/rules/sidecars/pet.yaml:158]: >I think this needs to be reflected in the selectors- i.e., we need two mutually exclusive sets of selectors. One with "none" in sidecar.ReconMethodParameterLabels and one with !("none" in sidecar.ReconMethodParameterLabels).

Are you suggesting the below?

```
RuleOne:
  selectors: 
      - "none" in sidecar.ReconMethodParameterLabels
  fields:
    ReconMethodParameterUnits: optional
    ReconMethodParameterValues: optional

RuleTwo:
   selectors:
       - !("none" in sidecar.ReconMethodParameterLabels)
   fields:
     ReconMethodParameterUnits: required
     ReconMethodParameterValues: required
```

Do we really need mutual exclusion for this? It seems excessive when one rule covers both cases, or is it a technical limitation with the validator or schema tools?

**Comment by effigies** [comment]
@bendhouseart While working through some examples, we've found that it's going to be very difficult to both render and enforce rules that have `required` fields and then relax under certain conditions. So we can do two things:

1) Make the optional default and use `level_addendum` to describe when it's required, and then add a required rule with a selector. Here only the first table will get rendered, and it will say something like `OPTIONAL, but REQUIRED if XYZ`.

```YAML
PETReconstruction:
  selectors:
    - modality == "pet"
    - suffix == "pet"
  fields:
    ReconMethodParameterLabels:
      level: required
    ReconMethodParameterUnits:
      level: optional
      level_addendum: |
        required if `ReconMethodParameterLabels` does not contain 'none'

PETReconstructionReq:
  selectors:
    - modality == "pet"
    - suffix == "pet"
    - '!intersects(sidecar.ReconMethodParameterLabels, ["none"])'
  fields:
    ReconMethodParameterUnits: required
```

2) Split the rule into two separate tables, one `required` and one `optional`, and then combine during the render:

```YAML
PETReconstruction:
  selectors:
    - modality == "pet"
    - suffix == "pet"
  fields:
    ReconMethodParameterLabels:
      level: required

PETReconstructionReq:
  selectors:
    - modality == "pet"
    - suffix == "pet"
    - '!intersects(sidecar.ReconMethodParameterLabels, ["none"])'
  fields:
    ReconMethodParameterUnits:
      level: required
      level_addendum: |
        optional if `ReconMethodParameterLabels` does not contain 'none'

PETReconstructionOpt:
  selectors:
    - modality == "pet"
    - suffix == "pet"
    - intersects(sidecar.ReconMethodParameterLabels, ["none"])
    ReconMethodParameterUnits: optional
```

```Markdown
{{ MACROS___make_sidecar_table(["pet.PETReconstruction". "pet.PETReconstructionReq"]) }}
```

**Comment by CPernet** [comment]
@effigies isn't `OPTIONAL but REQUIRED if ` similar to what is already used for ['NonlinearGradientCorrection'](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#sequence-specifics) which then would make sense to use a  rule for conditional stuff -- although that case this is `RECOMMENDED, but REQUIRED if ` (which would also actually be better for _ReconMethod Parameters_ and _ReconFilter Parameters_

**Comment by effigies** [comment]
Yes, these are equivalent approaches in terms of enforcement. The choice between the two is which way PET maintainers would prefer it to be rendered.

Note that if you make a field RECOMMENDED, there will be warnings when it is missing. (The current validator does not follow this rule, but the schematized one will.)

**Comment by melanieganz** [comment]
@effigies, so we just discussed this again in the OpenNeuroDev meeting and we agreed upon going for  `RECOMMENDED, but REQUIRED if ` . We'd like it to spit out a warning during validation , since we want to encourage people to go look for the values, but we accept that they might be hard to find/dig out.

File: src/schema/rules/sidecars/pet.yaml [hunk 5]
```diff
@@ -146,16 +146,20 @@ PETReconstruction:
       level: required
       description_addendum: This partly matches the DICOM Tag (0054,1103) `Reconstruction Method`.
     ReconMethodParameterUnits:
-      level: required
+      level: recommended
+      level_addendum: Recommended, but required if `ReconMethodParameterLabels` does not contain 'none'.
```
Comment (effigies) [inline src/schema/rules/sidecars/pet.yaml:150]:
```suggestion
      level_addendum: required if `ReconMethodParameterLabels` does not contain `"none"`
```

File: src/schema/rules/sidecars/pet.yaml [hunk 6]
```diff
@@ -146,16 +146,20 @@ PETReconstruction:
...
       description_addendum: This partly matches the DICOM Tag (0054,1103) `Reconstruction Method`.
     ReconMethodParameterValues:
-      level: required
+      level: recommended
+      level_addendum: |
+        Recommended, but required if `ReconMethodParameterLabels` does not contain 'none'.
```
Comment (effigies) [inline src/schema/rules/sidecars/pet.yaml:155]:
```suggestion
      level_addendum: required if `ReconMethodParameterLabels` does not contain `"none"`
```

File: src/schema/rules/sidecars/pet.yaml [hunk 7]
```diff
@@ -146,16 +146,20 @@ PETReconstruction:
...
       description_addendum: This partly matches the DICOM Tag (0054,1103) `Reconstruction Method`.
     ReconFilterType:
       level: required
       description_addendum: This partly matches the DICOM Tag (0018,1210) `Convolution Kernel`.
     ReconFilterSize:
-      level: required
+      level: recommended
+      level_addendum: Recommended, but required if `ReconFilterType` is not "none".
```
Comment (effigies) [inline src/schema/rules/sidecars/pet.yaml:162]:
```suggestion
      level_addendum: required if `ReconFilterType` is not `"none"`
```

**Review by effigies — APPROVED** [review]
Looks right. Some styling fixes for the table.

**Comment by effigies** [comment]
Final render: image

**Comment by effigies** [comment]
@melanieganz @CPernet @mnoergaard Any of you up for giving this another review and a green check if you're happy with it?

**Comment by sappelhoff** [comment]
tagging again: @melanieganz @CPernet @mnoergaard

**Comment by sappelhoff** [comment]
Thanks all!

**Pull request merged**